bigtable/client/polling_policy.h

```diff
@@ -0,0 +1,52 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_POLLING_POLICY_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_POLLING_POLICY_H_
+
+#include "bigtable/client/version.h"
+#include <grpc++/grpc++.h>
+#include <chrono>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Define the interface for providing async repetitive call rules
+ *
+ */
+class PollingPolicy {
```

Comment:
I think the PollingPolicy needs work.  Consider the following suggestions.  You may want to implement them in a separate PR:

The application needs to tell us for how long they are willing to wait, and how to backoff between retries.  They can tell us by giving us a number of retries or a total number of retries (or a combination of them), so this is similar to the `RPCRetryPolicy`.  We recommend using exponential backoff for these loops, so we should make it easy to use that, but possible to use something more hard-coded.

So maybe a `PollingPolicy` is a combination of a retry policy and a backoff policy?  I would start with something like this:

```C++
class PollingPolicy { public:
  virtual std::unique_ptr<PollingPolicy> clone() = 0;
  // Return true if `status` represents a permanent error that cannot be retried.
  virtual bool IsPermanentError(gprc::Status const& status) = 0;
  // Return true if we cannot try again.
  virtual bool Exhausted() = 0;
  // Return for how long we should wait before trying again.
  virtual std::chrono::milliseconds WaitPeriod() = 0;
};
```

The default polling policy should be a combination of the (existing) Retry and Backoff policies:

```C++
template <
   typename Retry = LimitedTimeRetryPolicy,
   typename Backoff = ExponentialBackoffPolicy>
class GenericPollingPolicy : public PollingPolicy, private Retry, private Backoff  { public:
  GenericPollingPolicy() : Retry(), Backoff() {}
  GenericPollingPolicy(Retry retry, Backoff backoff)
   : Retry(std::move(retry)), Backoff(std::move(backoff)) {}
  
  std::unique_ptr<PollingPolicy> clone() { return new GenericPollingPolicy(*this); }
  bool IsPermanentError(gprc::Status const& status) override {
    return not Retry::can_retry(status);
  }
  bool Exhausted() override {
    return not Retry::on_failure(grpc::Status::OK);
  }
  std::chrono::milliseconds WaitPeriod() {
    return Backoff::on_completion(grpc::Status::OK);
  }
};

std::unique_ptr<PollingPolicy> DefaultPollingPolicy() {
  return new GenericPollingPolicy<>();
}
```

I think that allows us to reuse the existing classes, and